src/platform_impl/windows/event_loop.rs

```diff
@@ -599,40 +601,91 @@ impl<T: 'static> EventLoopProxy<T> {
 
 type WaitUntilInstantBox = Box<Instant>;
 
+/// A lazily-initialized window message ID.
+pub struct LazyId {
+    /// The ID.
+    id: AtomicU32,
+
+    /// The name of the message.
+    name: &'static str,
+}
+
+/// An invalid custom window ID.
+const INVALID_ID: u32 = 0x0;
+
+impl LazyId {
+    /// Create a new `LazyId`.
+    const fn new(name: &'static str) -> Self {
+        Self {
+            id: AtomicU32::new(INVALID_ID),
+            name,
+        }
+    }
+
+    /// Get the message ID.
+    pub fn get(&self) -> u32 {
+        // Load the ID.
+        let mut id = self.id.load(Ordering::Relaxed);
+
+        // If the ID is invalid, create one.
+        if id == INVALID_ID {
+            // Register the message.
+            // SAFETY: We are sure that the pointer is a valid C string ending with '\0'.
+            assert!(self.name.ends_with('\0'));
+            let new_id = unsafe { RegisterWindowMessageA(self.name.as_ptr()) };
+
+            // The new ID should never be zero.
+            assert_ne!(
+                new_id,
+                0,
+                "RegisterWindowMessageA returned zero for '{}': {}",
+                self.name,
+                std::io::Error::last_os_error()
+            );
+
+            // Try to store the new ID.
+            let swapped_id = self
+                .id
+                .compare_exchange(id, new_id, Ordering::SeqCst, Ordering::SeqCst)
```

Comment:
I don't think sequential consistency is required here, `Relaxed` should be fine.

See [the relevant chapter](https://marabos.nl/atomics/atomics.html#example-racy-init) in Mara Bos' new book about atomics for details.

Reply:
I replaced it with `AcqRel` and `Acquire`. I'd like to avoid a potential race condition with another event loop.